### PR TITLE
Implement sam service — modular architecture

### PR DESCRIPTION
**Pull request by Nico-La-Porta**
Implement the SAM2 Model Service using FastAPI to enhance modularity and facilitate the integration of new models into the segmentation web application.

Features:
- Add model_service.py (FastAPI service).
- Introduce utilities for Points, Bboxes, Frames, and Masks.
- Add test for SamManager.

### CODE REVIEW

File: Dockerfile.model [hunk 1]
```diff
@@ -0,0 +1,72 @@
+# SAM2 Model Service Dockerfile
+# ARG BASE_IMAGE=pytorch/pytorch:2.4.1-cuda12.4-cudnn9-devel
+ARG BASE_IMAGE=pytorch/pytorch:2.3.1-cuda12.1-cudnn8-runtime
+ARG MODEL_SIZE=tiny
```
Comment (ChataingT) [inline Dockerfile.model:4]: The size of sam2 is decided here. Is it a default that may be overriden later ? 
If not we do not need to download all weights.
If yes, why does it need to be instanciated inside the docker as a global environnement variable ?

File: Dockerfile.model [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+# Install Python dependencies
+RUN pip install --upgrade pip setuptools
+# Install SAM 2 package in editable mode
+RUN pip install -e ".[dev]"
+
+# Install additional dependencies for the model service
+RUN pip install fastapi uvicorn requests
+
+# Expose port for model service
+EXPOSE 8001
```
Comment (ChataingT) [inline Dockerfile.model:65]: This expose the port 8001 but what happen if instead of the default command we use --port 8002 because the 8001 is occupied ?

Do we have to force the use of a specific port for this docker ? Are the other part of the application dependant of this port value ?

File: sam2/sam2_video_predictor.py [hunk 3]
```diff
@@ -9,11 +9,11 @@
 
 import torch
 import gc
-import cv2
+# import cv2
```
Comment (ChataingT) [inline sam2/sam2_video_predictor.py:12]: If unused, it is better to delete.

File: sam2/sam2_video_predictor.py [hunk 4]
```diff
@@ -9,11 +9,11 @@
...
 
 from sympy.physics.units import current
 from tqdm import tqdm
-import math
+# import math
```
Comment (ChataingT) [inline sam2/sam2_video_predictor.py:16]: If package unused, it is better to delete the line

File: sam2/sam_manager/sam_manager.py [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+import logging
+
+# Add SAM2 to path
+SAM2_PATH = os.path.join(os.path.dirname(os.path.dirname(os.path.dirname(os.path.abspath(__file__)))))
+sys.path.insert(0, SAM2_PATH)
+
+from sam2.build_sam import build_sam2_video_predictor
+
+from ..server.utils import *
+log = logging.getLogger('myapp')
```
Comment (ChataingT) [inline sam2/sam_manager/sam_manager.py:15]: Because this project is meant to be used with others, we can give it a specific name in the logging in case we aggregate all log later.

File: sam2/sam_manager/sam_manager.py [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+
+this_file_path = os.path.dirname(os.path.abspath(__file__))
+MODEL_CHECKPOINTS_DIR = os.getenv("MODEL_CHECKPOINTS_DIR", os.path.join(this_file_path, "../../checkpoints"))
+MODEL_CONFIG_DIR = os.getenv("MODEL_CONFIG_DIR", os.path.join(this_file_path, "../configs/sam2_1"))
+
+class SamManager():
```
Comment (ChataingT) [inline sam2/sam_manager/sam_manager.py:21]: Can you add some comment to explain the score of this class ?

File: sam2/sam_manager/sam_manager.py [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+            log.info("SAM2 state reset successfully")
+            return True
+
+
+if __name__ == "__main__":
+
+    samM = SamManager()
+    test_video_path = os.path.join(this_file_path, "../../test/assets/bedroom.mp4")
+    samM.init_model(test_video_path)
+
```
Comment (ChataingT) [inline sam2/sam_manager/sam_manager.py:289]: To clean the file this should be in the test directory.

File: sam2/server/model_service.py [hunk 8]
```diff
@@ -0,0 +1,200 @@
...
+sys.path.insert(0, SAM2_PATH)
+
+from sam2.sam_manager.sam_manager import SamManager
+from sam2.server.utils import FrameManager
+
+from model_requests import *
+
+# Configure logging
+logging.basicConfig(level=logging.INFO)
+logger = logging.getLogger(__name__)
```
Comment (ChataingT) [inline sam2/server/model_service.py:27]: The logger is not consistent with the logger used in the other files. We should have a global config and logger. If neededd, for debugging purpose, the log message can have the file of origins.

File: sam2/server/model_service.py [hunk 9]
```diff
@@ -0,0 +1,200 @@
...
+
+os.environ['HYDRA_FULL_ERROR'] = '1'
```
Comment (ChataingT) [inline sam2/server/model_service.py:29]: Is it a behavior we want to have in debug and production ?

File: sam2/server/model_service.py [hunk 10]
```diff
@@ -0,0 +1,200 @@
...
+
+# ThreadPoolExecutor for running tasks in the background
+executor = ThreadPoolExecutor(max_workers=1)
+
+this_file_path = os.path.dirname(os.path.abspath(__file__))
+OUTPUT_LOCAL = os.path.join(this_file_path, "../../data/output")
+INPUT_VIDEO_DIR = os.path.join(this_file_path, "../../data/input")
```
Comment (ChataingT) [inline sam2/server/model_service.py:36]: I assume those value are hardcoded to work inside the docker. Is it the expected behavior ?
If yes, it should at least be mentionned and alerted.
If not, those should be arguments.

File: sam2/server/model_service.py [hunk 11]
```diff
@@ -0,0 +1,200 @@
...
+        logger.info("SAM2 state reset successfully")
+        return JSONResponse(content={"message": "SAM2 state reset successfully"}, status_code=200)
+
+if __name__ == "__main__":
+    uvicorn.run(
+        app,
+        host="127.0.0.1",
+        port=8001,
+        log_level="info"
+    )
```
Comment (ChataingT) [inline sam2/server/model_service.py:200]: Those should be in test or in private file for debugging and local runs.

File: sam2/server/utils/bbox.py [hunk 12]
```diff
@@ -0,0 +1,26 @@
+class BBox():
```
Comment (ChataingT) [inline sam2/server/utils/bbox.py:1]: A documentation about the scope of the bbox will be a plus here.

File: sam2/server/utils/frame.py [hunk 13]
```diff
@@ -0,0 +1,90 @@
+from typing import List
+
+from .bbox import BBox
+from .point import Point
+from .mask import FrameMask
+
+class Frame():
```
Comment (ChataingT) [inline sam2/server/utils/frame.py:10]: Same about the documentation especiallw related to the other class

File: sam2/server/utils/mask.py [hunk 14]
```diff
@@ -0,0 +1,120 @@
+import numpy as np
+from typing import List, Dict, Any
+import torch
+
+class FrameMask():
```
Comment (ChataingT) [inline sam2/server/utils/mask.py:5]: Same doc

File: sam2/server/utils/point.py [hunk 15]
```diff
@@ -0,0 +1,29 @@
+class Point():
```
Comment (ChataingT) [inline sam2/server/utils/point.py:1]: Same doc

File: setup.py [hunk 16]
```diff
@@ -22,7 +22,10 @@
 
 # Required dependencies
 REQUIRED_PACKAGES = [
-    "torch>=2.4.1",
+    # "torch>=2.4.1",
+    # "torchvision>=0.18.1",
+    # "numpy>=1.26.0",
+    "torch>=2.3.1",
```
Comment (ChataingT) [inline setup.py:28]: Better to delete if unused but what bring the change in the torch version ?

File: test/run_tests.py
Comment (ChataingT) [inline test/run_tests.py:1]: Interesting test file

**Review by ChataingT — CHANGES_REQUESTED** [review]
Good implementation packaging sam2 in a server and in a docker.
It miss some documentation about class and about the whole. On how it should be used
The repo was first created as a fork of sam2 It should still be usable that way. Allowing for a docker version is a plus, same for the server but those should be independant. I mean that sam2 should be able to :
1. be installed and used alone
2. used inside a docker
3. used through an api running on the system
4. used as an api running inside a docker

File: sam2/sam_manager/sam_manager.py (hunk 6 above)
Comment (Nico-La-Porta) [reply sam2/sam_manager/sam_manager.py:21]: The scope of the class is to make it possible to use of this version of SAM with a modular logic (for the container-based app).
